### DISCUSSION TITLE
Refresh token is not valid.  Unable to authenticate with Ring servers

### DISCUSSION THREAD

**Discussion by cardsfan7189** [dgreif/ring#1402] [Q&A] [answered]
I was able to obtain the refresh token through the ring-auth-cli, but not sure I'm copying it to the right file. I tried copying it to a config.json in the same project directory where api-example.js is located.  Later copied it to an .env file in the same directory (probably ignorant on my part), but in both cases get this 
```
C:\Users\xxxxx\typescript_projects\ring> node .\api-example.js
Error: Refresh token is not valid.  Unable to authenticate with Ring servers.  See https://github.com/dgreif/ring/wiki/Refresh-Tokens
    at RingRestClient.getGrantData (C:\Users\xxxxx\typescript_projects\node_modules\ring-client-api\lib\rest-client.js:140:15)
    at RingRestClient.getAuth (C:\Users\xxxxx\typescript_projects\node_modules\ring-client-api\lib\rest-client.js:143:32)
    at get authPromise [as authPromise] (C:\Users\xxxxx\typescript_projects\node_modules\ring-client-api\lib\rest-client.js:90:38)
    at RingRestClient.request (C:\Users\xxxxx\typescript_projects\node_modules\ring-client-api\lib\rest-client.js:287:50)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async RingApi.fetchRawLocations (C:\Users\xxxxx\typescript_projects\node_modules\ring-client-api\lib\api.js:205:50)
    at async RingApi.fetchAndBuildLocations (C:\Users\xxxxx\typescript_projects\node_modules\ring-client-api\lib\api.js:217:30)
    at async example (C:\Users\xxxxx\typescript_projects\ring\api-example.js:9:21)
```
Tried searching this forum and others via Google but could not find a solution

EDIT:  added some logging to the api-example script and it's an environment issue.  the refresh token env variable is undefined, so working on that now

**Comment by tsightler** [comment, marked as answer]
The examples all use [dotenv](https://github.com/motdotla/dotenv) to read the configuration from a .env file.  You can find a full example project [here](https://github.com/dgreif/ring-client-example) which includes instructions on how to set the .env file with the token.  These are simple examples, you can choose to store and update the token however you like.

**Reply by cardsfan7189** [reply]
Thanks! I eventually found the ring-client-example and realized I wasn't formatting the .env file correctly.  Got it working, but hadn't returned to this discussion to post an answer.  I'll mark your reply as the answer. Now just need to find out why ring camera events are not being captured, but at least now can debug and try out different methods.

**Comment by tsightler** [comment]
> Now just need to find out why ring camera events are not being captured, but at least now can debug and try out different methods.

This is almost always caused by not saving the initial updated token, but instead trying to use the original token.  Listening for onRefreshTokenUpdated events and saving the new token is absolutely critical, especially for the very initial connection, otherwise, you will find that push notification only work the very first time.  Once an initial auth token is used, if you didn't save the updated token that is immediately returned on initial connection, push notification will not work again until you delete the device from Ring Control Center, recreate a new token, and save the token correctly the next time.

It seems most users do not understand the criticality of this step.  The reason this is important is because the ring-auth-cli tool only generates the refresh token needed for connecting to Ring API, however, the first time you actually use the token to make a connection to the Ring API, the code also generates a device registration to receive Firebase push notifications and re-encodes this data into the updated refresh token.  This **MUST BE SAVED** at this point, if it is not, future attempts to connect to the Ring API will not be able to receive push notifications.
